Add support for declaring metadata-only dependencies

nmeum opened the pull request:
Presently, it is not possible to only depend on the metadata of an item. This is problematic because it can cause a lot of unnecessary rebuilds when a compiler only needs to run on metadata changes. For example, consider a navigation sidebar that includes all tags of available posts. If this sidebar is included on every page, then we would always rebuild everything when changing the content of any post (even though the metadata tags themselves didn't change!).

To overcome this limitation, we need to distinguish between content and metadata dependencies. This is basically what #383 is about, where @jaspervdj also expressed interest in this feature. However, since nobody has worked on this [since 2015](https://github.com/jaspervdj/hakyll/issues/383#issuecomment-150836917), I decided to give it a shot.

The implementation proposed here is split into two commits, which can be reused independently. The first makes `outOfDate` capable of classifying Identifiers for which only the metadata changed. Building upon that, the second commit refactors `Hakyll.Core.Dependencies` (API-breaking change!) to introduce a `DependencyKind` which allows us to distinguish `KindContent` and `KindMetadata`. The commit also then makes use of the feature where applicable (`getRoute`, `compilerGetMetadata`, `buildTagsWith`). <!--There is potential to make more use of the feature, for example, providing a `loadMetadata` in `Hakyll.Core.Compiler.Require` but that presupposes agreeing on the new `Hakyll.Core.Dependencies` API and can probably be done separately.-->

Feedback and additional testing would be much appreciated! 🤗

---

Fixes #383

nmeum committed:
Refactor the Dependency type to allow metadata-only dependencies

This is an API-breaking change, it refactors the old Dependency type
into three types: A Dependency, a DependencyKind, and a
DependencySelector. Through the DependencyKind we can specify a
dependency on either the Metadata or the Content of another Identifier.

The DependencySelector allows us, for both DependencyKinds, to either
depend on a single Identifier or a Pattern (as before). Dependencies of
KindMetadata are only considered out-of-date if the metadata changed.

This is extremely useful in many scenarios where Hakyll previously
induced unnecessary rebuilds because it could not distinguish between a
content and a metadata dependency. As an example, consider a navigation
sidebar displaying all available tags. There are also examples in the
existing code base where I have now used a metadata-only dependency
(getRoute, compilerGetMetadata, buildTagsWith).

nmeum committed:
Provider: Classify identifiers for which only the metadata changed

This is the foundation work required to introduce dependencies only on
metadata. Often times, a post's body/content may change, but its
metadata remains unmodified. Some compilers only need to be re-run if
the metadata changes (e.g., consider a sidebar that includes all post
tags). Presently, Hakyll does not allow depending on metadata, thereby
incurring a lot of unnecessary in such a scenario.

This commit allows classification of identifiers for which only the
metadata changed. The next commit will add an API to only depend on the
metadata of an identifier.

In technical terms it is implemented as follows: When invalidating the
metadataCache, retrieve the old metadata values. Afterward, for
identifiers modified (according to resourceModified), load the new
metadata and compare it to the old. Only consider metadata to have
changed if it differs.

LaurentRDC commented on lib/Hakyll/Core/Dependencies.hs:
```diff
@@ -88,6 +106,19 @@ markOod id' = State.modify $ \s ->
+      0 -> pure KindContent
+      1 -> pure KindMetadata
+      _ -> error "Data.Binary.get: Invalid DependencyKind"
+
```
How do we ensure that the binary cache is clearly broken by this addition, and does not remain incoherent but still parseable?

One way would be to replace the tokens `0` and `1` by something more explicit, but larger, such as `"kindContent"` and `"kindMetadata"`, but maybe I'm being overly cautious.

LaurentRDC commented on lib/Hakyll/Core/Dependencies.hs:
```diff
@@ -88,6 +106,19 @@ markOod id' = State.modify $ \s ->
+  get = getWord8 >>= \t -> case t of
+      0 -> pure KindContent
+      1 -> pure KindMetadata
+      _ -> error "Data.Binary.get: Invalid DependencyKind"
```
Might as well switch to `fail` here as well

LaurentRDC commented on lib/Hakyll/Core/Dependencies.hs:
```diff
@@ -30,23 +34,37 @@ import           Hakyll.Core.Identifier.Pattern
+    get = getWord8 >>= \t -> case t of
+      0 -> Dependency <$> get <*> get
+      1 -> pure AlwaysOutOfDate
+      _ -> error "Data.Binary.get: Invalid Dependency"
```
`error` -> `fail`

LaurentRDC reviewed (commented):
Thanks for your contribution! I've got a few comments, no showstoppers.

nmeum commented on lib/Hakyll/Core/Dependencies.hs:
```diff
@@ -88,6 +106,19 @@ markOod id' = State.modify $ \s ->
+
+  get = getWord8 >>= \t -> case t of
+      0 -> pure KindContent
+      1 -> pure KindMetadata
```
I wasn't sure how compatibility wrt. the binary cache is handled across versions. I assumed it was recommended to do a clean build after an upgrade anyhow. However, keeping backwards compatibility should be rather trivial if we do something along the lines of:

```Haskell
instance Binary Dependency where
    put (Dependency k s) = putWord8 0 >> put k >> put s
    put AlwaysOutOfDate = putWord8 1

    get = getWord8 >>= \t -> case t of
      -- XXX: Backwards compatability with Hakyll <=4.16.7.1.
      0 -> (\p i -> Dependency KindContent $ PatternDependency p i) <$> get <*> get
      1 -> Dependency KindContent . IdentifierDependency <$> get

      2 -> pure AlwaysOutOfDate
      3 -> Dependency <$> get <*> get
      _ -> fail "Data.Binary.get: Invalid Dependency"
```

Then `0`, `1` (and `3`) will be parsed in the old format and written to disk in the new format. WDYT?

LaurentRDC commented:
There are more changes than last time I took a look. Was this the result of testing @nmeum ?

nmeum commented:
No functional changes, I just fixed some code style issues I noticed while reading the PR myself again, added some Haddock comments, rebased and squashed the fixup commits.

LaurentRDC reviewed (approved):
Thanks for your contribution!

nmeum commented:
For those interested in an extended usage example for this feature, I [wrote a tutorial](https://notes.8pit.net/notes/dijr.html) explain how one can generate [a sidebar](https://notes.8pit.net/notes/dijr.html#generating-a-sidebar) and [tag pages](https://notes.8pit.net/notes/dijr.html#generating-tag-pages) using metadata-only dependencies.

UnlimitedCookies commented:
I may be wrong, but it seems that with this patch something like
```
import Hakyll
[...]
    scssDependencies <- makePatternDependency "css/**.scss"
    rulesExtraDependencies [scssDependencies] $ do
      create ["css/main.scss"] $ do
        route $ setExtension "css"
        compile compressScssCompiler
[...]
```
now looks like this
```
import Hakyll
import Hakyll.Core.Dependencies
[...]
    scssDependencies <- makePatternDependency KindContent "css/**.scss"
    rulesExtraDependencies [scssDependencies] $ do
      create ["css/main.scss"] $ do
        route $ setExtension "css"
        compile compressScssCompiler
[...]
```
So beforehand all that was needed to define a dependency was the prelude, but now the dependency type also needs to be explicitly imported.

Is this intended or am I mistaking?

nmeum commented:
I guess I just never noticed because all of my source files already had `Hakyll.Core.Dependencies` imported. Could you propose a PR where you re-export `Dependency(..)`, `contentDependency`, and `metadataDependency` from the `Haykll` module?
